irohad/ametsuchi/mutable_storage.hpp

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_MUTABLESTORAGE_HPP
+#define IROHA_MUTABLESTORAGE_HPP
+
+#include <ametsuchi/command_executor.hpp>
+
+namespace iroha {
+  namespace ametsuchi {
+    /**
+     * Mutable storage is used apply blocks to the storage.
+     * Allows to query the world state view, transactions, and blocks.
+     */
+    class MutableStorage : public WsvQuery, public BlockQuery {
+      /**
+       * Applies a block to current mutable state
+       * using logic specified in function
+       * @param block Block to be applied
+       * @param function Function that specifies the logic used to apply the
+       * block
+       * Function parameters:
+       *  - Block @see block
+       *  - CommandExecutor
+       *  - WsvQuery
+       * Function returns true if the block is successfully applied, false
+       * otherwise.
+       * @return True if block was successfully applied, false otherwise.
+       *
+       */
+      virtual bool apply(
+          dao::Block block,
+          std::function<bool(dao::Block, CommandExecutor, WsvQuery)>
```

Comment:
What feature of this `std::function`?
I think this function does:
1. stateful validation. If validation is failed, returns false.
2. executor. Check the content of tx and apply new data to WSV.